neural_lam/models/ar_model.py

```diff
@@ -21,35 +22,67 @@ class ARModel(pl.LightningModule):
-
-        # Load static features for grid/data
-        static_data_dict = utils.load_static_data(
-            self.config_loader.dataset.name
+        self._datastore = datastore
+        # XXX: should be this be somewhere else?
+        split = "train"
+        num_state_vars = datastore.get_num_data_vars(category="state")
+        num_forcing_vars = datastore.get_num_data_vars(category="forcing")
+        da_static_features = datastore.get_dataarray(
+            category="static", split=split
         )
-        for static_data_name, static_data_tensor in static_data_dict.items():
-            self.register_buffer(
-                static_data_name, static_data_tensor, persistent=False
-            )
+        da_state_stats = datastore.get_normalization_dataarray(category="state")
+        da_boundary_mask = datastore.boundary_mask
+
+        # Load static features for grid/data, NB: self.predict_step assumes
+        # dimension order to be (grid_index, static_feature)
+        arr_static = da_static_features.transpose(
+            "grid_index", "static_feature"
+        ).values
+        self.register_buffer(
+            "grid_static_features",
+            torch.tensor(arr_static, dtype=torch.float32),
+            persistent=False,
+        )
+
+        state_stats = {
+            "state_mean": torch.tensor(
+                da_state_stats.state_mean.values, dtype=torch.float32
+            ),
+            "state_std": torch.tensor(
+                da_state_stats.state_std.values, dtype=torch.float32
+            ),
+            "diff_mean": torch.tensor(
+                da_state_stats.state_diff_mean.values, dtype=torch.float32
+            ),
+            "diff_std": torch.tensor(
+                da_state_stats.state_diff_std.values, dtype=torch.float32
+            ),
+        }
+
+        for key, val in state_stats.items():
+            self.register_buffer(key, val, persistent=False)
 
         # Double grid output dim. to also output std.-dev.
         self.output_std = bool(args.output_std)
         if self.output_std:
             # Pred. dim. in grid cell
-            self.grid_output_dim = 2 * self.config_loader.num_data_vars()
+            self.grid_output_dim = 2 * num_state_vars
         else:
             # Pred. dim. in grid cell
-            self.grid_output_dim = self.config_loader.num_data_vars()
+            self.grid_output_dim = num_state_vars
             # Store constant per-variable std.-dev. weighting
-            # Note that this is the inverse of the multiplicative weighting
+            # NOTE that this is the inverse of the multiplicative weighting
             # in wMSE/wMAE
+            # TODO: Do we need param_weights for this?
             self.register_buffer(
                 "per_var_std",
-                self.step_diff_std / torch.sqrt(self.param_weights),
+                self.diff_std,
                 persistent=False,
             )
```

Comment:
Yes, we need the variable weights here. Do I understand correctly that this PR currently removes all computation and use of such weights?

Handling and computations of these weights is something we should rework, but there is no need to do that in this PR. We should however not get rid of them completely, as they are needed to train any model. I would be fine with doing something simple for now, like just specifying all these weights manually in the yaml config. Then we can look into better way to compute and include these later.

Reply:
Yes, as I read @sadamov's multizarr branch it was removed. I am not sure exactly where this should sit though. It doesn't quite make sense to me to put it in the preparation of input data (i.e. generating the input zarr). Maybe this should be in the general training config? We don't currently have a config specification for that and I'm hesitant to add more in this PR. Can we maybe add a placeholder tensor/ndarray of length matching the number of features with a placeholder comment to say this should be moved to a config?

Reply:
I agree that this is not abvious, and it should probably not be computed as part of preparing the input data.

In the long term I think that we would want a system where you specify a "strategy" for this weighting, and the exact weighting is then computed from the input variables (taking e.g. their pressure level into account) using this strategy. One such strategy could just be to specify each weight manually in the config file.

For now I think we could just add a single entry in the general training config that is just one long list of all these weights. Then we can rework it later. This does not require any involved config specification, just a simple entry. The reason I would prefer that as a temporary solution over e.g. hard-coding a tensor of 1s is that it at least leaves the repo in a fully useable state. It's just a bit annoying to specify your variable weights.

Reply:
I am bit confused about the naming here (`per_var_std`) and therefore confused about what this is used for. As I read it `self.step_diff_std` is actually the *per-feature standard deviation of **difference** between consecutive timesteps* where the name `per_var_std` suggests to me that that variable is the *per-feature standard deviation*. I dont' think both of these can be correct. Also, I don't think it makes sense for the per-feature weighting here. As I am reading https://github.com/mllam/neural-lam/blob/main/neural_lam/models/ar_model.py#L172 this variable (as in `per_var_std`) is returned as the `per-variable std dev` when the model isn't trained to predict the std dev. But the per-variable std dev shouldn't change based on the weighting we apply I don't think, or am I missing something here?

Reply:
I am inferring that `self.step_diff_std` is the *per-feature standard deviation of the difference between consecutive timesteps* by following the following currently on `master`:

- assigned from load from disk in: https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/models/ar_model.py#L34
- loaded from disk in: https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/utils.py#L62
- written to disk by: https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/create_parameter_weights.py#L92
- written to disk by call from: https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/create_parameter_weights.py#L385

Reply:
And just to clarify the explicit question also:

> But the per-variable std dev shouldn't change based on the weighting we apply I don't think, or am I missing something here?

It does here, as all forms of weighting are incorporated into a predicted std-dev. This is the way to interpret (possibly shoe-horn a little) a weighted MSE loss function as maximum-likelihood estimation of a probabilistic model.

Reply:
Ok, I understand the difference between `self.step_diff_std` vs `self.per_var_std` now, that makes sense. Thanks! Although I still think it is confusing to refer to `self.per_var_std` as being a *prediction*, true it is used as a prediction by the model, but as it is derived from static content it is rather a static estimate of the per-variable standard deviation per variable. But at least we are on the same page about this, that's great!

> It contains the corresponding std-dev for each variable. When this specific constant is used in the weighted MSE loss as
> 
> https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/metrics.py#L77

This I'm still unsure about. You say this computes a weighted loss, but I don't see a weight in the expression. Instead I would expect to see something like `error * feature_based_weight / feature_variance_scaling`. The "weighting" that we provide by hand (like the higher weight being placed on near-surface variables) should be a separate term to me, because the `pred_std` is no longer the standard deviation, but rather that scaled somehow. Or am I missing the point of how this should work?

I have introduced the config-idea that we discussed elsewhere on this issue where I have introduced to option to define per-feature-weights and made an example config here: https://github.com/leifdenby/neural-lam/commit/b57bc7ac0c10d9467f04f2500a078543c9b310a1#diff-8bef849b5dfe26a4f672f721b4120adc43de1f50585e6460ab2a8d23c215269c. This is missing actually using these weights because I'm not sure where they should go in. I was expected to see the weighting values used in the loss-function calculation, or maybe to weight the errors before they go into the loss.

Reply:
> Although I still think it is confusing to refer to self.per_var_std as being a prediction
Yea, that's fair. It's mostly just convenient to think always of it as that, since that will be the case when `self.output_std=True`.

> This I'm still unsure about. You say this computes a weighted loss, but I don't see a weight in the expression. Instead I would expect to see something like `error * feature_based_weight / feature_variance_scaling`.

The pred_std takes the role of the weight.  If you insert `pred_std = sqrt(feature_variance_scaling / feature_based_weight)`, which is exactly what is saved as `per_var_std` at https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/models/ar_model.py#L50-L54 , into https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/metrics.py#L77 you will get `entry_mse * feature_based_weight / feature_variance_scaling` just as you expect.

At the end of the day this is just a simple way to encompass all variable weighting into std-dev predictions from the model. This then matches the probabilistic interpretations of such weightings. But this might not be the best way to do this, and you might prefer pushing all weighting into metrics.py, rather than combining it into a weighting in the form of a std.-dev. in ARModel, which is done now. Anyhow I don't think such a change has anything to do with this PR, so better to not change that here.

The config with `state_feature_weights` looks great! Earlier the weights came from https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/utils.py#L69-L73 but I guess now you could do something like

```python
self.param_weights = datastore.get_state_feature_weights()
```
in the ARModel constructor, before 
https://github.com/mllam/neural-lam/blob/a2ddcd4298d7865225c45866e7495664b3e382ab/neural_lam/models/ar_model.py#L50-L54.

I guess you would have to implement `datastore.get_state_feature_weights` then and make it return a torch tensor with the correct variable ordering.